lib/src/kbasesearchengine/main/IndexerWorker.java

```diff
@@ -865,28 +950,45 @@ public ObjectTypeParsingRules getTypeDescriptor(final SearchObjectType type)
         }
         
         @Override
-        public Map<GUID, SearchObjectType> getTypesForGuids(Set<GUID> guids)
+        public Map<GUID, SearchObjectType> getTypesForGuids(final List<GUID> objectRefPath, final Set<GUID> unresolvedGUIDs)
                 throws InterruptedException, IndexingException {
             Map<GUID, SearchObjectType> ret = new LinkedHashMap<>();
-            Set<GUID> guidsToLoad = new LinkedHashSet<>();
-            for (GUID guid : guids) {
+            Set<ResolvedReference> guidsToLoad = new LinkedHashSet<>();
+            for (GUID unresolvedGUID : unresolvedGUIDs) {
+                ResolvedReference resRef =
```

Comment:
That would be an optimization that is based on the current use of the method. If the use changes, for instance by sending in a mixed bag of guids, this optimization will become a breaking change. For this reason, I would rather leave it as is.

Reply:
Not sure I'm following. I mean something like this:

```java
final Set<RR> rrs = resolveRefs(objectRefPath, unresolvedGUIDs);
for (final RR resRef: rrs) {
    final GUID guid = resRef.getResolvedReference();
...
```

That's exactly the same as the current implemenation except there's 1 call to resolveRefs vs. N.

Reply:
OK, I spent an hour trying to figure out why I did it this way after spending an hour figuring the same thing out the last time I did it this way. The reason is because the subObjectType and subObjectId gets lost in the resolveReference process and the only way I can get it is by iterating through the unresolved guids and then getting this info from the unresolved guid in each iteration. 

The reason why this info is lost is because the cache used in the line below keys on a ref path that includes only parent ref info (1/2/1;1/3/1 for example). So by the time the sub-objects get resolved, they get the parent info and lose all sub-object info. 

https://github.com/arfathpasha/KBaseSearchEngine/blob/master/lib/src/kbasesearchengine/main/IndexerWorker.java#L738

Reply:
I see. In that case, IIUC, you could do

```java
final Set<RR> rrs = resolveRefs(objectRefPath, unresolvedGUIDs);
final Map<GUID, RR> rrmap = rrs.stream().toMap(rr -> rr.getReference(), rr);
for (final GUID unresolvedGUID: unresolvedGUIDS) {
    final RR resRef = rrmap.get(unresolvedGUID.parentGUID());
    ...
```

Not sure if I have all the syntax and method calls right but you get the idea.

Reply:
Done.